File: node/ts/zkgroup/calllinks/CallLinkAuthCredentialResponse.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { randomBytes } from 'node:crypto';

import ByteArray from '../internal/ByteArray.js';
import * as Native from '../../Native.js';
import { RANDOM_LENGTH } from '../internal/Constants.js';

import GenericServerSecretParams from '../GenericServerSecretParams.js';
import GenericServerPublicParams from '../GenericServerPublicParams.js';
import CallLinkAuthCredential from './CallLinkAuthCredential.js';
import { Aci } from '../../Address.js';

export default class CallLinkAuthCredentialResponse extends ByteArray {
  private readonly __type?: never;

  constructor(contents: Uint8Array) {
    super(contents, Native.CallLinkAuthCredentialResponse_CheckValidContents);
  }

  static issueCredential(
    userId: Aci,
    redemptionTime: number,
    params: GenericServerSecretParams
  ): CallLinkAuthCredentialResponse {
    const random = randomBytes(RANDOM_LENGTH);
    return this.issueCredentialWithRandom(
      userId,
      redemptionTime,
      params,
      random
    );
  }

  static issueCredentialWithRandom(
    userId: Aci,
    redemptionTime: number,
    params: GenericServerSecretParams,
    random: Uint8Array
  ): CallLinkAuthCredentialResponse {
    return new CallLinkAuthCredentialResponse(
      Native.CallLinkAuthCredentialResponse_IssueDeterministic(
        userId.getServiceIdFixedWidthBinary(),
        redemptionTime,
        params.contents,
        random
      )
    );
  }

  receive(
    userId: Aci,
    redemptionTime: number,
    params: GenericServerPublicParams
  ): CallLinkAuthCredential {
    return new CallLinkAuthCredential(
      Native.CallLinkAuthCredentialResponse_Receive(
        this.contents,
        userId.getServiceIdFixedWidthBinary(),
        redemptionTime,
        params.contents
      )
    );
  }
}

File: java/shared/java/org/signal/libsignal/zkgroup/calllinks/CallLinkAuthCredentialResponse.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.calllinks;

import org.signal.libsignal.protocol.ServiceId.Aci;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.VerificationFailedException;
import org.signal.libsignal.zkgroup.GenericServerPublicParams;
import org.signal.libsignal.zkgroup.GenericServerSecretParams;
import org.signal.libsignal.zkgroup.internal.ByteArray;
import org.signal.libsignal.internal.Native;

import java.security.SecureRandom;
import java.time.Instant;

import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

public final class CallLinkAuthCredentialResponse extends ByteArray {
  public CallLinkAuthCredentialResponse(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.CallLinkAuthCredentialResponse_CheckValidContents(contents);
  }

  public static CallLinkAuthCredentialResponse issueCredential(Aci userId, Instant redemptionTime, GenericServerSecretParams params) {
    return issueCredential(userId, redemptionTime, params, new SecureRandom());
  }

  public static CallLinkAuthCredentialResponse issueCredential(Aci userId, Instant redemptionTime, GenericServerSecretParams params, SecureRandom secureRandom) {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.CallLinkAuthCredentialResponse_IssueDeterministic(userId.toServiceIdFixedWidthBinary(), redemptionTime.getEpochSecond(), params.getInternalContentsForJNI(), random);

    try {
      return new CallLinkAuthCredentialResponse(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public CallLinkAuthCredential receive(Aci userId, Instant redemptionTime, GenericServerPublicParams params) throws VerificationFailedException {
    byte[] newContents = Native.CallLinkAuthCredentialResponse_Receive(getInternalContentsForJNI(), userId.toServiceIdFixedWidthBinary(), redemptionTime.getEpochSecond(), params.getInternalContentsForJNI());

    try {
      return new CallLinkAuthCredential(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}
